File: docs/samples/hello.py
"""Greetings in a few languages, with enough Python in it to show a colour theme."""
from __future__ import annotations

import sys
from dataclasses import dataclass
from enum import Enum
from functools import cached_property

DEFAULT_TARGET = "World"
MAX_WIDTH = 42


class Language(Enum):
    ENGLISH = "en"
    GREEK = "el"
    FRENCH = "fr"


@dataclass(frozen=True)
class Greeting:
    language: Language
    template: str

    def render(self, target: str) -> str:
        return self.template.format(target=target)


class Greeter:
    """Formats greetings and keeps count of how many it has produced."""

    def __init__(self, greetings: list[Greeting], width: int = MAX_WIDTH) -> None:
        self._greetings = greetings
        self._width = width
        self.count = 0

    @cached_property
    def languages(self) -> set[Language]:
        return {greeting.language for greeting in self._greetings}

    def greet(self, target: str = DEFAULT_TARGET) -> list[str]:
        lines = [greeting.render(target) for greeting in self._greetings]
        self.count += len(lines)
        if any(len(line) > self._width for line in lines):
            raise ValueError(f"a greeting exceeds {self._width} characters")
        return lines


def main(argv: list[str]) -> int:
    target = argv[1] if len(argv) > 1 else DEFAULT_TARGET
    greeter = Greeter([
        Greeting(Language.ENGLISH, "Hello, {target}!"),
        Greeting(Language.GREEK, "Γεια σου, {target}!"),
        Greeting(Language.FRENCH, "Bonjour, {target} !"),
    ])
    for index, line in enumerate(greeter.greet(target), start=1):
        print(f"{index:>2}. {line}")
    print(f"-- {greeter.count} greetings in {len(greeter.languages)} languages")
    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv))

File: docs/samples/hello.cpp
// Greetings in a few languages, with enough C++ in it to show a colour theme.
#include <cstdint>
#include <iostream>
#include <string>
#include <string_view>
#include <vector>

namespace crawler {

constexpr std::string_view kDefaultTarget = "World";
constexpr std::size_t kMaxWidth = 42;

enum class Language : std::uint8_t { English, Greek, French };

struct Greeting {
    Language language;
    std::string_view prefix;
    std::string_view suffix;

    [[nodiscard]] std::string render(std::string_view target) const {
        return std::string{prefix} + std::string{target} + std::string{suffix};
    }
};

class Greeter {
public:
    explicit Greeter(std::vector<Greeting> greetings, std::size_t width = kMaxWidth)
        : greetings_{std::move(greetings)}, width_{width} {}

    template <typename Sink>
    void greet(std::string_view target, Sink&& sink) {
        for (const auto& greeting : greetings_) {
            const std::string line = greeting.render(target);
            if (line.size() > width_) {
                throw std::length_error("a greeting exceeds the maximum width");
            }
            sink(++count_, line);
        }
    }

    [[nodiscard]] int count() const noexcept { return count_; }

private:
    std::vector<Greeting> greetings_;
    std::size_t width_;
    int count_ = 0;
};

}  // namespace crawler

int main(int argc, char** argv) {
    const std::string_view target = argc > 1 ? argv[1] : crawler::kDefaultTarget;
    crawler::Greeter greeter{{
        {crawler::Language::English, "Hello, ", "!"},
        {crawler::Language::Greek, "Γεια σου, ", "!"},
        {crawler::Language::French, "Bonjour, ", " !"},
    }};
    greeter.greet(target, [](int index, const std::string& line) {
        std::cout << index << ". " << line << '\n';
    });
    std::cout << "-- " << greeter.count() << " greetings\n";
    return 0;
}
